### ISSUE TITLE
False positive when UUID is used as key in dict or set

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5513]
With the recent change made for #5446, pyright now incorrectly reports an error when a `uuid.UUID` instance is used as a key because it thinks it's not hashable. This is because of the logic recently added for #5446. It could be fixed in typeshed by adding an explicit `__hash__` method to the `UUID` class definition. I'm worried that there are other class definitions in typeshed that also define `__eq__` but omit a `__hash__` even though a `__hash__` is included in the implementation.

Here's a minimal repro:
```python
from uuid import UUID

def func(key: UUID):
    return {key: 0} # Error: UUID is not hashable
```

@TheTripleV, I'm half tempted to revert the change for #5446 because I'm worried it's going to create a bunch of false positives. I'd much rather have a few edge-case false negatives than false positives.

If you want to keep #5446 in place, I'd like to ask you to:
1. Fix the UUID issue in typeshed.
2. Think about whether there's a way we can find other cases in typeshed that are similar. I don't know if this can be done in an automated way. It may require a manual review of all classes in typeshed that claim to implement `__eq__` but not `__hash__`.

**Comment by AlexWaygood** [comment]
I'm up for trying to fix this in typeshed if it will improve pyright's inference. We can probably modify [stubtest](https://mypy.readthedocs.io/en/stable/stubtest.html) (which we use in typeshed's CI) so that it flags when `__hash__` is overridden at runtime but not in the stub. I've provided lots of patches to stubtest before, and would be happy to work on this.

In the short term, however, I also think it would be reasonable for you to revert https://github.com/microsoft/pyright/issues/5446 if you think it would cause too many false positives, while we figure out how many `__hash__` methods might be missing from typeshed

**Comment by AlexWaygood** [comment]
According to a stubtest patch I just wrote, around 40 or so classes in typeshed's stdlib stubs have `__hash__` overridden with a non-`None` value at runtime, but not in the stub:

<details>

```
error: builtins.frozenset.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\builtins.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'frozenset' objects>

error: builtins.memoryview.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\builtins.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'memoryview' objects>

error: builtins.range.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\builtins.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'range' objects>

error: builtins.tuple.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\builtins.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'tuple' objects>

error: collections.UserString.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\collections\__init__.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\collections\__init__.py:1354
<function UserString.__hash__ at 0x000001E5EA981760>

error: contextvars.ContextVar.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\contextvars.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of '_contextvars.ContextVar' objects>

error: datetime.datetime.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\datetime.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'datetime.datetime' objects>

error: enum.Enum.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\enum.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\enum.py:1229
<function Enum.__hash__ at 0x000001E5EA9D13A0>

error: importlib.machinery.ExtensionFileLoader.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\importlib\machinery.pyi
MISSING
Runtime: in file <frozen importlib._bootstrap_external>
<function ExtensionFileLoader.__hash__ at 0x000001E5EA6F2F20>

error: importlib.metadata.EntryPoint.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\importlib\metadata\__init__.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\importlib\metadata\__init__.py:277
<function EntryPoint.__hash__ at 0x000001E5F70E1B20>

error: inspect.Parameter.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\inspect.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\inspect.py:2795
<function Parameter.__hash__ at 0x000001E5EB370040>

error: inspect.Signature.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\inspect.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\inspect.py:3063
<function Signature.__hash__ at 0x000001E5EB370AE0>

error: ipaddress.IPv4Interface.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\ipaddress.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\ipaddress.py:1435
<function IPv4Interface.__hash__ at 0x000001E5F996DC60>

error: ipaddress.IPv6Address.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\ipaddress.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\ipaddress.py:1933
<function IPv6Address.__hash__ at 0x000001E5F996EC00>

error: ipaddress.IPv6Interface.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\ipaddress.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\ipaddress.py:2140
<function IPv6Interface.__hash__ at 0x000001E5F996F9C0>

error: ipaddress._BaseNetwork.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\ipaddress.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\ipaddress.py:736
<function _BaseNetwork.__hash__ at 0x000001E5F998FBA0>

error: pstats.FunctionProfile.__mypy-replace is not present at runtime
Stub: in file stdlib\pstats.pyi:32
def (*, ncalls: builtins.str =, tottime: builtins.float =, percall_tottime: builtins.float =, cumtime: builtins.float =, percall_cumtime: builtins.float =, file_name: builtins.str =, line_number: builtins.int =)
Runtime:
MISSING

error: pstats.FunctionProfile.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\pstats.pyi
MISSING
Runtime: in file <string>
<function FunctionProfile.__hash__ at 0x000001E5FB0A0CC0>

error: pstats.StatsProfile.__mypy-replace is not present at runtime
Stub: in file stdlib\pstats.pyi:41
def (*, total_tt: builtins.float =, func_profiles: builtins.dict[builtins.str, pstats.FunctionProfile] =)
Runtime:
MISSING

error: pstats.StatsProfile.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\pstats.pyi
MISSING
Runtime: in file <string>
<function StatsProfile.__hash__ at 0x000001E5FB0A1080>

error: re.Pattern.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\re.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 're.Pattern' objects>

error: tracemalloc.Frame.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\tracemalloc.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\tracemalloc.py:169
<function Frame.__hash__ at 0x000001E5F9A6BE20>

error: tracemalloc.Statistic.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\tracemalloc.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\tracemalloc.py:42
<function Statistic.__hash__ at 0x000001E5F9984AE0>

error: tracemalloc.StatisticDiff.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\tracemalloc.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\tracemalloc.py:84
<function StatisticDiff.__hash__ at 0x000001E5F99845E0>

error: tracemalloc.Trace.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\tracemalloc.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\tracemalloc.py:300
<function Trace.__hash__ at 0x000001E5F9A6B380>

error: tracemalloc.Traceback.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\tracemalloc.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\tracemalloc.py:212
<function Traceback.__hash__ at 0x000001E5F9A6BBA0>

error: types.BuiltinFunctionType.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\types.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'builtin_function_or_method' objects>

error: types.BuiltinMethodType.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\types.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'builtin_function_or_method' objects>

error: types.CodeType.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\types.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'code' objects>

error: types.GenericAlias.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\types.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'types.GenericAlias' objects>

error: types.MethodType.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\types.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'method' objects>

error: types.MethodWrapperType.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\types.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'method-wrapper' objects>

error: types.UnionType.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\types.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'types.UnionType' objects>

error: unittest.FunctionTestCase.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\unittest\__init__.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\unittest\case.py:1427
<function FunctionTestCase.__hash__ at 0x000001E5ED4BA0C0>

error: unittest.TestCase.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\unittest\__init__.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\unittest\case.py:509
<function TestCase.__hash__ at 0x000001E5ED3F32E0>

error: unittest.case.FunctionTestCase.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\unittest\case.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\unittest\case.py:1427
<function FunctionTestCase.__hash__ at 0x000001E5ED4BA0C0>

error: unittest.case.TestCase.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\unittest\case.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\unittest\case.py:509
<function TestCase.__hash__ at 0x000001E5ED3F32E0>

error: uuid.UUID.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\uuid.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\uuid.py:268
<function UUID.__hash__ at 0x000001E5F5FC8AE0>

error: weakref.WeakMethod.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\weakref.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'weakref.ReferenceType' objects>

error: winreg.HKEYType.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\winreg.pyi
MISSING
Runtime:
<slot wrapper '__hash__' of 'PyHKEY' objects>

error: xml.etree.ElementTree.QName.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\xml\etree\ElementTree.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\xml\etree\ElementTree.py:499
<function QName.__hash__ at 0x000001E5F66162A0>

error: xml.etree.cElementTree.QName.__hash__ overrides __hash__ at runtime but not in the stub
Stub: in file stdlib\xml\etree\cElementTree.pyi
MISSING
Runtime: in file C:\Users\alexw\AppData\Local\Programs\Python\Python311\Lib\xml\etree\ElementTree.py:499
<function QName.__hash__ at 0x000001E5F66162A0>

Found 42 errors (checked 541 modules)
```

</details>

Of these, it looks like around half of them currently override `__eq__`, which would presumably mean that pyright would currently view them as unhashable.

**Comment by erictraut** [comment]
Thanks @AlexWaygood, that's incredibly helpful. Based on this, I'm going to revert the change for #5446 for the time being. We can revisit it later if/when these issues are addressed in typeshed.

**Comment by erictraut** [comment]
Closing this issue because it's no longer a problem with the other change reverted.

**Comment by AlexWaygood** [comment]
https://github.com/python/typeshed/commit/03b4bb9ccecbabff7ebd7f98bd5c5239f2a33b97 added many missing `__eq__` and `__hash__` methods to typeshed's stdlib stubs (including `uuid.UUID.__hash__`).

I've yet to run my analysis on typeshed's third-party stubs (which is a little more fiddly), but will try to do so in the next few days. Once that's been done, I can submit my stubtest patch as a PR to mypy, so that missing `__hash__` and `__eq__` overrides will be flagged in typeshed's CI.

**Comment by erictraut** [comment]
Thanks @AlexWaygood. That eliminates some of the false positives generated by the expanded hashability check proposed by @TheTripleV, but `mypy_primer` indicates that the check still results in false positives in other popular third-party libraries such as `bidict` and `pandas-stubs`. For that reason, I'm going to leave it reverted for now.

@TheTripleV, if you want to work with the authors of `bidict` and `pandas-stubs` to update their stubs accordingly, let me know once they've released updates.

**Comment by TheTripleV** [comment]
@erictraut Fixes in both repos have been merged. The bidict false positive was only in test cases, not in the library. pandas-stubs has cut a new releases.

**Comment by erictraut** [comment]
OK, thanks for pursing those changes. The [mypy-primer results ](https://github.com/microsoft/pyright/pull/5723)are clean(er) now. I've un-reverted the earlier change, and I'll include this in the next pyright release.
